File: monsoon3.0/src/main/java/wtf/monsoon/impl/ui/recode/panel/elements/Panel.java
package wtf.monsoon.impl.ui.recode.panel.elements;

import lombok.NonNull;
import lombok.Setter;
import me.surge.animation.Animation;
import me.surge.animation.Easing;
import org.lwjgl.nanovg.NanoVG;
import wtf.monsoon.Wrapper;
import wtf.monsoon.api.module.Category;
import wtf.monsoon.api.ui.Comp;
import wtf.monsoon.api.util.render.ColorUtil;

import java.awt.*;
import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicInteger;

public class Panel extends Comp {
    @Setter
    float veloX = 0;

    Category category;
    public Animation expanded = new Animation(() -> 300F, false, () -> Easing.CIRC_IN_OUT);
    ArrayList<ModuleComp> moduleComps = new ArrayList<>();
    public Panel(Category category, @NonNull float x, @NonNull float y, @NonNull float w, @NonNull float h) {
        super(x, y, w, h);
        this.category = category;
        AtomicInteger o = new AtomicInteger();
        Wrapper.getMonsoon().getModuleManager().getModulesByCategory(category).forEach(m -> {
            moduleComps.add(new ModuleComp(m,x,y+h+h* o.get(),w,h));
            o.getAndIncrement();
        });
    }

    @Override
    public void render(float mx, float my) {
//        ui.dropShadow(x,y,w,2,1,5,ColorUtil.interpolate(ColorUtil.getAccent()[0], ColorUtil.TRANSPARENT,0.5), ColorUtil.TRANSPARENT,true);
//        ui.dropShadow(x,y,w,2,1,10,ColorUtil.interpolate(ColorUtil.getAccent()[0], ColorUtil.TRANSPARENT,0.7), ColorUtil.TRANSPARENT,true);
//
//        ui.dropShadow(x,y+h+h*moduleComps.size()*(float)expanded.getAnimationFactor(),w,2,1,5,ColorUtil.interpolate(ColorUtil.getAccent()[0], ColorUtil.TRANSPARENT,0.5), ColorUtil.TRANSPARENT,true);
//        ui.dropShadow(x,y+h+h*moduleComps.size()*(float)expanded.getAnimationFactor(),w,2,1,10,ColorUtil.interpolate(ColorUtil.getAccent()[0], ColorUtil.TRANSPARENT,0.7), ColorUtil.TRANSPARENT,true);

        float maxHeight = 0;
        for (ModuleComp moduleComp : moduleComps) {
            maxHeight += moduleComp.getAbsoluteHeight();
        }

        ui.finishFrame();
        ui.initFrame();
        NanoVG.nvgTranslate(ui.vg,mx,my);
        NanoVG.nvgRotate(ui.vg,veloX);
        NanoVG.nvgTranslate(ui.vg,-(mx),-(my));

//        ui.dropShadow(x+8,y+8,w-16,h-16+maxHeight*(float) expanded.getAnimationFactor(),0,10,ColorUtil.getAccent()[0], ColorUtil.TRANSPARENT,false);

        ui.rect(x,y,w,h+h*moduleComps.size()*(float) expanded.getAnimationFactor(), new Color(0x161616));
        ui.rect(x,y,w,h, new Color(0x101010));
        ui.text(category.getIcon(), "category2", 20, x+8, y+h/2f,Color.WHITE, 16);
        ui.roundedLinearGradient(x,y,w,2, 0,ColorUtil.getAccent()[0],ColorUtil.getAccent()[1]);
        ui.text(category.name(), "sbold", 16, x+w/2f, y+h/2f+1, Color.WHITE, 2 | 16);


        ui.rect(x, y + h, w, maxHeight * (float) expanded.getAnimationFactor(), new Color(0x161616));

        if(expanded.getAnimationFactor() > 0.001) {
            ui.scissor(x, y + h, w, maxHeight * (float) expanded.getAnimationFactor(), () -> {
                float[] i = {0};
                moduleComps.forEach(m -> {
                    m.setX(x);
                    m.setY(y + h + i[0]);
                    m.render(mx, my);
                    i[0]+=m.getAbsoluteHeight();
                });
            });
        }
        ui.roundedLinearGradient(x,y+h+maxHeight*(float)expanded.getAnimationFactor(),w,2, 0,ColorUtil.getAccent()[0],ColorUtil.getAccent()[1]);
    }

    @Override
    public void click(float mx, float my, int button) {
        if(expanded.getState())
            moduleComps.forEach(m -> m.click(mx, my, button));
    }

    @Override
    public void key(int code, char c) {
        if(expanded.getState())
            moduleComps.forEach(m -> m.key(code,c));
    }
}

File: monsoon-3.x-v2/src/main/java/wtf/monsoon/client/ui/kpanel/element/Panel.kt
package wtf.monsoon.client.ui.kpanel.element

import me.surge.animation.Animation
import me.surge.animation.Easing
import wtf.monsoon.Wrapper
import wtf.monsoon.backend.Category
import wtf.monsoon.client.util.ui.NVGWrapper
import java.awt.Color

/**
 * @author surge
 * @since 21/02/2023
 */
class Panel(val category: Category, x: Float, y: Float, width: Float, height: Float) : Element(x, y, width, height) {

    private val expanded = Animation(200f, true, Easing.CIRC_IN_OUT)
    private val children = arrayListOf<ModuleElement>()

    var dragging = false

    var velocity: Float = 0f

    private var lastX = 0f
    private var lastY = 0f

    private var lastRenderX = x
    private var lastRenderY = y

    init {
        Wrapper.monsoon.moduleManager.getModuleByCategory(category).forEach { module ->
            children.add(ModuleElement(module, x, y, width - 4, height))
        }
    }

    override fun draw(mouseX: Int, mouseY: Int, mouseDelta: Float) {
        if (dragging) {
            this.x = mouseX - lastX
            this.y = mouseY - lastY

            velocity = (mouseX - lastX - lastRenderX) / 360
        } else {
            velocity = 0f
        }

        ui.endFrame()
        ui.beginFrame()

        ui.save()

        ui.translate(mouseX.toFloat(), mouseY.toFloat())
        ui.rotate(velocity)
        ui.translate(-mouseX.toFloat(), -mouseY.toFloat())

        ui.rect(x, y, width, trueHeight(), Color(16, 16, 16))
        ui.rect(x, y, width, height, Color(10, 10, 10))
        ui.text(category.iconCode.toString(), x + 8, y + height / 2f, "category2", 20f, Color.WHITE, NVGWrapper.Alignment.LEFT_MIDDLE)
        ui.roundedLinearGradient(x, y, width, 2f, 0f, Color.BLUE, Color(161, 0, 255))
        ui.text(category.name, x + width / 2f, y + height / 2f + 1, "sbold", 16f, Color.WHITE, NVGWrapper.Alignment.CENTER_MIDDLE)

        ui.rect(x, y + height, width, (offset() * expanded.getAnimationFactor()).toFloat(), Color(16, 16, 16))

        if (expanded.getAnimationFactor() > 0) {
            ui.scissor(x, y + height, width, (offset() * expanded.getAnimationFactor()).toFloat()) {
                var offset = 0f

                children.forEach { child ->
                    child.x = x + 2
                    child.y = y + height + offset

                    child.draw(mouseX, mouseY, mouseDelta)

                    offset += child.trueHeight()
                }
            }
        }

        // ui.roundedLinearGradient(x, (y + height + offset() * expanded.getAnimationFactor()).toFloat(), width, 2f, 0f, Color.BLUE, Color(161, 0, 255))

        ui.restore()

        lastRenderX = x
        lastRenderY = y
    }

    override fun click(mouseX: Int, mouseY: Int, button: Int): Boolean {
        if (hovered(mouseX, mouseY)) {
            when (button) {
                0 -> {
                    lastX = mouseX - x
                    lastY = mouseY - y

                    dragging = true
                }

                1 -> {
                    expanded.state = !expanded.state
                }
            }
        }

        if (expanded.state) {
            children.forEach { child ->
                if (child.click(mouseX, mouseY, button)) {
                    return true
                }
            }
        }

        return super.click(mouseX, mouseY, button)
    }

    override fun release(mouseX: Int, mouseY: Int, button: Int): Boolean {
        if (button == 0) {
            this.dragging = false
        }

        if (expanded.state) {
            children.forEach { child ->
                if (child.release(mouseX, mouseY, button)) {
                    return true
                }
            }
        }

        return super.release(mouseX, mouseY, button)
    }

    override fun key(code: Int, char: Char): Boolean {
        if (expanded.state) {
            children.forEach { child ->
                if (child.key(code, char)) {
                    return true
                }
            }
        }

        return super.key(code, char)
    }

    override fun offset(): Float {
        return (children.sumOf { it.trueHeight().toDouble() } * expanded.getAnimationFactor()).toFloat()
    }

}